src/main/java/com/artipie/SliceFromConfig.java

```diff
@@ -0,0 +1,130 @@
+import com.artipie.files.FilesSlice;
+import com.artipie.gem.GemSlice;
+import com.artipie.http.Response;
+import com.artipie.http.Slice;
+import com.artipie.maven.http.MavenSlice;
+import com.artipie.npm.Npm;
+import com.artipie.npm.http.NpmSlice;
+import com.artipie.rpm.http.RpmSlice;
+import com.jcabi.log.Logger;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import org.reactivestreams.Publisher;
+
+/**
+ * Slice from repo config.
+ * @since 0.1.4
+ * @todo #90:30min We still don't have tests for Pie. But now that this class was extracted, we have
+ *  a more cohesive class that could be tested. Write unit tests for SliceFromConfig class.
+ */
+public final class SliceFromConfig implements Slice {
+
+    /**
+     * Repository config.
+     */
+    private final RepoConfig config;
+
+    /**
+     * Ctor.
+     * @param config Repo config
+     */
+    public SliceFromConfig(final RepoConfig config) {
+        this.config = config;
+    }
+
+    @Override
+    public Response response(final String line, final Iterable<Map.Entry<String, String>> headers,
+        final Publisher<ByteBuffer> body) {
+        try {
+            return SliceFromConfig.build(this.config).response(
+                line, headers, body
+            );
+        } catch (final InterruptedException ex) {
+            Logger.error(this, "Interruption when getting slice from config");
+            throw new IllegalArgumentException(ex);
+        } catch (final ExecutionException ex) {
+            Logger.error(this, "Exception when getting slice from config");
+            throw new IllegalArgumentException(ex);
+        }
+    }
+
+    /**
+     * Find a slice implementation for config.
+     * @param cfg Repository config
+     * @return Slice
+     * @throws ExecutionException If error getting the slice
+     * @throws InterruptedException If error getting the slice
+     */
+    private static Slice build(final RepoConfig cfg) throws InterruptedException,
+        ExecutionException {
+        return cfg.type().thenCombine(
+            cfg.storage(),
+            (type, storage) -> {
+                final Slice slice;
+                switch (type) {
+                    case "file":
+                        slice = new FilesSlice(storage);
+                        break;
+                    case "npm":
+                        slice = new NpmSlice(new Npm(storage), storage);
+                        break;
+                    case "gem":
+                        slice = new GemSlice(storage);
+                        break;
+                    case "rpm":
+                        slice = new RpmSlice(storage);
+                        break;
+                    case "php":
+                        try {
+                            slice = cfg.path().thenApply(
+                                path -> new PhpComposer(path, storage)
+                            ).toCompletableFuture().get();
```

Comment:
@HDouss I dont think we should execute the request here, it should return `CompletitionStage`

Reply:
@Vatavuk it is not executing the request, it is resolving config.path and build the slice. This method is returning a slice

Reply:
@HDouss We are fetching yaml content from a repo in an asynchronous manner. The code you've wrote will be synchronous.

Reply:
@Vatavuk Actually it is not really synchronous. Because all this method is called when the `response` method of the slice is called. What is really synchronous is the call to the constructor of `SliceFromConfig` which is much less than the older switch/case logic (that was actually synchronous)

Reply:
@HDouss you are right, I didn't catch the execution of the methotd itself